devito/passes/clusters/cse.py

```diff
@@ -156,53 +160,59 @@ def _compact_temporaries(exprs, exclude):
 
 
 @singledispatch
-def count(expr):
+def count(expr, conds):
     """
     Construct a mapper `expr -> #occurrences` for each sub-expression in `expr`.
     """
     mapper = Counter()
     for a in expr.args:
-        mapper.update(count(a))
+        mapper.update(count(a, None))
     return mapper
 
 
 @count.register(list)
 @count.register(tuple)
-def _(exprs):
+def _(exprs, conds):
     mapper = Counter()
     for e in exprs:
-        mapper.update(count(e))
+        mapper.update(count(e, None))
     return mapper
 
 
+@count.register(ClusterizedEq)
+def _(exprs, conds):
+    conditionals = exprs.conditionals
+    return count(exprs.rhs, conditionals)
```

Comment:
I'll write it here, but it applies to the whole dispatching logic.

I don't think you need to propagate `conds` through the various `count` recursive calls.

In particular here `count` returns a dict; all I think you have to do here is:

```
mapper = count(expr.rhs)
if expr.conditionals:
     mapper = {(k, expr.conditionals): v for k, v in mapper.items()}
return mapper
```

To be more precise, we shouldn't just use a 2-tuple as key, but rather either a namedtuple or a Bunch, so that the keys are homogeneous and you don't have to do `k[0]` at the call site

Reply:
`namedtuple` is probably OK, but you could also introduce a `Candidate` class such that `conditionals` in an optional argument; or preset the `namedtuple` so that `conditionals` defaults to None.

I don't know if one such `Candidate` class would help elsewhere, for example we could enrich it with properties such as `estimate` and `free_symbols` to decentralize complexity, though admittedly this is getting close to nitpicking

Anyway, I think it's good to keep the complexity of CSE as under control as possible

Reply:
```
mapper = count(expr.rhs)
if expr.conditionals:
     mapper = 
```

Thqt would require a by hand merge for list/..., propagating conditionals makes the Counter update much easier and simpler

I'm fine with a `namedtuple` and will fix, `Candidate` class would have to bee too intricated to handle different use case but long term can be looked into.